File: java_implementation/HammingCode.java
import java.util.Scanner;

public class HammingCode {
    // Method to calculate the number of parity bits required
    private static int calculateParityBits(int dataBits) {
        int parityBits = 0;
        while (Math.pow(2, parityBits) < (dataBits + parityBits + 1)) {
            parityBits++;
        }
        return parityBits;
    }

    // Method to generate the Hamming code
    public static String generateHammingCode(String data) {
        int dataBits = data.length();
        int parityBits = calculateParityBits(dataBits);
        int totalBits = dataBits + parityBits;
        char[] hammingCode = new char[totalBits];

        // Initialize the hamming code array with placeholders
        for (int i = 0; i < totalBits; i++) {
            hammingCode[i] = '0';
        }

        // Place data bits and parity bits at appropriate positions
        for (int i = 0, j = 0; i < totalBits; i++) {
            if ((i + 1 & i) == 0) {
                // Position is a power of 2, leave it for parity bit
                hammingCode[i] = '0';
            } else {
                // Position is for data bit
                hammingCode[i] = data.charAt(j);
                j++;
            }
        }

        // Calculate parity bits
        for (int i = 0; i < parityBits; i++) {
            int parityPos = (int) Math.pow(2, i) - 1;
            int parity = 0;
            for (int j = parityPos; j < totalBits; j += 2 * (parityPos + 1)) {
                for (int k = j; k < j + parityPos + 1 && k < totalBits; k++) {
                    if (k != parityPos) {
                        parity ^= Character.getNumericValue(hammingCode[k]);
                    }
                }
            }
            hammingCode[parityPos] = (char) (parity + '0');
        }

        return new String(hammingCode);
    }

    // Method to detect and correct a single-bit error in the received Hamming code
    public static String detectAndCorrectError(String received) {
        int totalBits = received.length();
        int parityBits = calculateParityBits(totalBits);
        int errorPos = 0;

        // Check parity bits
        for (int i = 0; i < parityBits; i++) {
            int parityPos = (int) Math.pow(2, i) - 1;
            int parity = 0;
            for (int j = parityPos; j < totalBits; j += 2 * (parityPos + 1)) {
                for (int k = j; k < j + parityPos + 1 && k < totalBits; k++) {
                    parity ^= Character.getNumericValue(received.charAt(k));
                }
            }
            if (parity != 0) {
                errorPos += parityPos + 1;
            }
        }

        // Correct error if found
        if (errorPos != 0) {
            System.out.println("Error detected at position: " + errorPos);
            char[] receivedArray = received.toCharArray();
            receivedArray[errorPos - 1] = receivedArray[errorPos - 1] == '0' ? '1' : '0';
            received = new String(receivedArray);
        } else {
            System.out.println("No error detected.");
        }

        return received;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Enter the data to be transmitted:");
        String data = scanner.nextLine();

        String hammingCode = generateHammingCode(data);
        System.out.println("Generated Hamming Code: " + hammingCode);

        System.out.println("Enter the received Hamming Code:");
        String received = scanner.nextLine();

        String correctedCode = detectAndCorrectError(received);
        System.out.println("Corrected Hamming Code: " + correctedCode);

        scanner.close();
    }
}

File: python_implementation/HammingCode.py
def calculate_parity_bits(data_bits):
    """Calculate the number of parity bits needed for the given number of data bits."""
    parity_bits = 0
    while (2 ** parity_bits) < (data_bits + parity_bits + 1):
        parity_bits += 1
    return parity_bits

def generate_hamming_code(data):
    """Generate the Hamming Code for the given data."""
    data_bits = len(data)
    parity_bits = calculate_parity_bits(data_bits)
    total_bits = data_bits + parity_bits
    hamming_code = ['0'] * total_bits

    # Place data bits and leave space for parity bits
    j = 0
    for i in range(1, total_bits + 1):
        if (i & (i - 1)) == 0:
            continue
        hamming_code[i - 1] = data[j]
        j += 1

    # Calculate parity bits
    for i in range(parity_bits):
        parity_pos = (2 ** i) - 1
        parity = 0
        for j in range(parity_pos, total_bits, 2 * (parity_pos + 1)):
            for k in range(j, min(j + parity_pos + 1, total_bits)):
                parity ^= int(hamming_code[k])
        hamming_code[parity_pos] = str(parity)

    return ''.join(hamming_code)

def detect_and_correct_error(received):
    """Detect and correct a single-bit error in the received Hamming Code."""
    total_bits = len(received)
    parity_bits = calculate_parity_bits(total_bits)
    error_pos = 0

    # Check parity bits
    for i in range(parity_bits):
        parity_pos = (2 ** i) - 1
        parity = 0
        for j in range(parity_pos, total_bits, 2 * (parity_pos + 1)):
            for k in range(j, min(j + parity_pos + 1, total_bits)):
                parity ^= int(received[k])
        if parity != 0:
            error_pos += parity_pos + 1

    # Correct error if found
    if error_pos != 0:
        print(f"Error detected at position: {error_pos}")
        received_list = list(received)
        received_list[error_pos - 1] = '1' if received_list[error_pos - 1] == '0' else '0'
        received = ''.join(received_list)
    else:
        print("No error detected.")

    return received

if __name__ == "__main__":
    data = input("Enter the data to be transmitted: ")
    hamming_code = generate_hamming_code(data)
    print("Generated Hamming Code: " + hamming_code)

    received = input("Enter the received Hamming Code: ")
    corrected_code = detect_and_correct_error(received)
    print("Corrected Hamming Code: " + corrected_code)
